components/sync15/src/engine/legacy_engine.rs

```diff
@@ -0,0 +1,189 @@
+//! mechanical to migrate a `LegacySyncEngine` to a `SyncEngine`.
+
+use super::sync_engine::{EngineSyncAssociation, SyncEngine};
+use super::CollectionRequest;
+use crate::bso::{IncomingBso, OutgoingBso};
+use crate::client_types::ClientData;
+use crate::{telemetry, CollectionName, Guid, ServerTimestamp};
+use anyhow::Result;
+use std::cell::{Cell, RefCell};
+
+/// A Changeset is a concept that only exists for the LegacySyncEngine.
+// Incoming and Outgoing changesets are almost identical except for the timestamp.
+#[derive(Debug)]
+pub struct IncomingChangeset {
+    pub changes: Vec<IncomingBso>,
+    /// The server timestamp of the collection.
+    pub timestamp: ServerTimestamp,
+    pub collection: CollectionName,
+}
+
+impl IncomingChangeset {
+    #[inline]
+    pub fn new(collection: CollectionName, timestamp: ServerTimestamp) -> Self {
+        Self::new_with_changes(collection, timestamp, Vec::new())
+    }
+
+    #[inline]
+    pub fn new_with_changes(
+        collection: CollectionName,
+        timestamp: ServerTimestamp,
+        changes: Vec<IncomingBso>,
+    ) -> Self {
+        Self {
+            changes,
+            timestamp,
+            collection,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct OutgoingChangeset {
+    pub changes: Vec<OutgoingBso>,
+    pub collection: CollectionName,
+}
+
+impl OutgoingChangeset {
+    #[inline]
+    pub fn new(collection: CollectionName, changes: Vec<OutgoingBso>) -> Self {
+        Self {
+            collection,
+            changes,
+        }
+    }
+}
+
+/// A LegacySyncEngine is a deprecated trait that reflects how we used to
+/// perform a sync. It can be thought of an an "adaptor" between old engines
+/// and the new world order.
+///
+/// We should try and kill this by moving all engines to the new trait. The priority
+/// should be engines which stage incoming and outgoing records to a database to
+/// avoid holding the records in memory.
+pub trait LegacySyncEngine {
+    fn collection_name(&self) -> CollectionName;
+    fn prepare_for_sync(&self, _get_client_data: &dyn Fn() -> ClientData) -> Result<()> {
+        Ok(())
+    }
+    fn set_local_encryption_key(&mut self, _key: &str) -> Result<()> {
+        unimplemented!("This engine does not support local encryption");
+    }
+    fn apply_incoming(
+        &self,
+        inbound: Vec<IncomingChangeset>,
+        telem: &mut telemetry::Engine,
+    ) -> Result<OutgoingChangeset>;
+    fn sync_finished(
+        &self,
+        new_timestamp: ServerTimestamp,
+        records_synced: Vec<Guid>,
+    ) -> Result<()>;
+    fn get_collection_requests(
+        &self,
+        server_timestamp: ServerTimestamp,
+    ) -> Result<Vec<CollectionRequest>>;
+    fn get_sync_assoc(&self) -> Result<EngineSyncAssociation>;
+    fn reset(&self, assoc: &EngineSyncAssociation) -> Result<()>;
+    fn wipe(&self) -> Result<()>;
+
+    // the magic
+    fn get_legacy_engine_state(&self) -> &LegacySyncEngineState;
```

Comment:
What if this returns a `&mut`, since we're going to be storing the staged records in the state?  Would that remove the need for the `RefCell<>` wrappers?

Reply:
that turns into this function needing a `&mut self` which explodes into everything needing it. We could juggle where the `RefCell`s are, but I wonder if we should lean into the engine traits taking `&mut self`? None of this stuff needs shared references. I might play with that in a different PR :)